### PR TITLE
Add linter rule for templated label PromQL

### PR DESCRIPTION
**Pull request by arajkumar**
This commit introduces basic support to lint promql expressions in template
queries.

Signed-off-by: Arunprasad Rajkumar <arajkuma@redhat.com>

### CODE REVIEW

File: lint/rule_template_label_promql.go [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+			}
+			for _, template := range d.Templating.List {
+				if template.Type == "query" {
+					if expr, err := parser.ParseExpr(template.Query); expr == nil {
+						return Result{
+							Severity: Error,
+							Message:  fmt.Sprintf("Dashboard '%s', template '%s' invalid PromQL query '%s': %v", d.Title, template.Name, template.Query, err),
+						}
+					}
+				}
```
Comment (tomwilkie) [inline lint/rule_template_label_promql.go:29]:
```suggestion
				if template.Type != "query" {
					continue
				}
				if expr, err := parser.ParseExpr(template.Query); expr == nil {
					return Result{
						Severity: Error,
						Message:  fmt.Sprintf("Dashboard '%s', template '%s' invalid PromQL query '%s': %v", d.Title, template.Name, template.Query, err),
					}
				}
```

Lets try and straight-line the code as much as possible.

**Review by tomwilkie — COMMENTED** [review]
This looks fantastic, thank you @arajkumar! Couple of minor comments, but thank you soo much for your contribution.

File: lint/rule_template_label_promql_test.go [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+				}{
+					List: []Template{
+						{
+							Type:  "datasource",
+							Query: "prometheus",
+						},
+						{
+							Name:       "namespaces",
+							Datasource: "$datasource",
+							Query:      "label_values(up{job=~\"$job\"})",
```
[the remark this answers is not in the document]
Comment (arajkumar) [reply lint/rule_template_label_promql_test.go:55]: @tomwilkie Thanks for reviewing the PR. 

Indeed this throws an error, however `parser.Expr` still had a valid `parser.Node` even with `label_values`. But the same becomes `nil` for an invalid promql expression. Pretty strange, but works for this case. I thought of using regex initially, but bit lazy to study on that as I'm not good at using regex. Let me apply your suggestion and improve it for this case.

File: lint/rule_template_label_promql.go [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+		if name == n {
+			return true
+		}
+	}
+	return false
+}
+
+// parseTemplatedLabelPromQL returns the parsed PromQL statement from a templated label,
+// also ensures that the label function name is valid.
+func parseTemplatedLabelPromQL(t Template) (parser.Expr, error) {
```
Comment (tomwilkie) [inline lint/rule_template_label_promql.go:25]: As we don't used the returned expr, lets not return it.

File: lint/rule_template_label_promql_test.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (tomwilkie) [reply lint/rule_template_label_promql_test.go]: Oh interesting I didn't know that!

**Comment by tomwilkie** [comment]
Looks great @arajkumar! Just one super minor nit then we can get this merged.  Thank you!

**Review by rgeyer — APPROVED** [review]
Late to the party, but this looks great. Thanks you very much for the contribution!

**Comment by tomwilkie** [comment]
Thank you @arajkumar!

**Comment by arajkumar** [comment]
Thank you @tomwilkie & @rgeyer for your kind words and time.

**Pull request merged**